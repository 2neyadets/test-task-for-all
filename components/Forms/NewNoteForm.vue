<template lang="pug">
  .new-note-form
    h6.mt-none.mb-md Создание новой заметки
    form(@submit.prevent="createNote" style="min-width: 270px;")
      MyInput.mb-lg(
        v-model.trim="form.title"
        label="Название"
        :maxlength="20"
        clearable
        autofocus
      )
      .flex.justify-end
        MyBtn(
          @click.native="$parent.$emit('hide')"
          label="Закрыть"
          size="16"
          flat
        )
        MyBtn.ml-sm(
          type="submit"
          label="Создать"
          :disable="!form.title"
          size="16"
        )
</template>

<script>
import { mapGetters } from 'vuex'
import MyInput from '../Reusable/MyInput'
import MyBtn from '../Reusable/MyBtn'
import { getId } from '../../utils/helpers'

const initialForm = () => ({
  title: null
})

export default {
  name: 'NewNoteForm',
  components: { MyBtn, MyInput },
  props: {},
  data () {
    return {
      form: initialForm(),
    }
  },
  computed: {
    ...mapGetters('notes', [
      'notesArr',
    ]),
  },
  watch: {
    'form.title' (v) {
      sessionStorage.setItem('createNoteForm.title', v)
    },
  },
  created () {},
  mounted () {
    const title = sessionStorage.getItem('createNoteForm.title')
    if (title !== 'null') this.form.title = title || null
  },
  updated () {},
  methods: {
    createNote () {
      const notesArr = [...this.notesArr]
      notesArr.push({ id: getId(), ...this.form })
      this.$store.dispatch('notes/changeNotes', notesArr)
      this.form = { ...this.form, ...initialForm() }
      sessionStorage.removeItem('createNoteForm.title')
      this.$parent.$emit('hide')
    },
  },
}
</script>

<style scoped lang="stylus">
</style>
