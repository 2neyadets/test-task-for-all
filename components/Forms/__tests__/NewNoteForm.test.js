import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils'
import NewNoteForm from '../NewNoteForm'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('NewNoteForm component', () => {
  let wrapper, actions, noteGetters, store

  beforeEach(() => {
    actions = {
      'notes/changeNotes': jest.fn(),
    }
    noteGetters = {
      notesArr: () => [],
    }
    store = new Vuex.Store({
      actions,
      modules: {
        notes: {
          namespaced: true,
          getters: noteGetters,
        },
      },
    })

    wrapper = shallowMount(NewNoteForm, {
      localVue,
      store,
    })
  })

  it('mounts without errors', done => {
    localVue.config.errorHandler = done // для отлавливания ошибок внутри nextTick, если нужно что-то выпонять внутри него
    expect.assertions(1)
    expect(wrapper).toBeTruthy()
    done()
  })

  it('can submit on btn', async () => {
    expect.assertions(1)
    wrapper.setData({
      form: {
        title: 'asdasdasd',
      },
    })
    await localVue.nextTick()
    const submitBtn = wrapper.findAll('MyBtn-stub').at(1)
    submitBtn.trigger('submit')
    expect(actions['notes/changeNotes']).toHaveBeenCalled()
  })

  it('can submit on input', async () => {
    expect.assertions(1)
    wrapper.setData({
      form: {
        title: 'asdasdasd',
      },
    })
    await localVue.nextTick()
    const myInput = wrapper.find('MyInput-stub')
    myInput.trigger('submit')
    expect(actions['notes/changeNotes']).toHaveBeenCalled()
  })

  it('match snapshot', () => {
    expect.assertions(1)
    expect(wrapper.element).toMatchSnapshot()
  })
})
