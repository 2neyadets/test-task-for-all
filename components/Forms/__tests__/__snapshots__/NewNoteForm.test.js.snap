// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewNoteForm component match snapshot 1`] = `
<div
  class="new-note-form"
>
  <h6
    class="mt-none mb-md"
  >
    Создание новой заметки
  </h6>
  <form
    style="min-width: 270px;"
  >
    <myinput-stub
      autofocus="true"
      class="mb-lg"
      clearable="true"
      hint=""
      label="Название"
      maxlength="20"
      mockid=""
    />
    <div
      class="flex justify-end"
    >
      <mybtn-stub
        flat="true"
        icon=""
        label="Закрыть"
        size="16"
        type="button"
      />
      <mybtn-stub
        class="ml-sm"
        disable="true"
        icon=""
        label="Создать"
        size="16"
        type="submit"
      />
    </div>
  </form>
</div>
`;
